### ISSUE TITLE
Irregular spacing in === and !== ligatures in Sublime Text on Windows

### ISSUE THREAD

**Issue by Steve-OH** [MonoLisaFont/feedback#131]
We welcome any feedback regarding technical issues (broken letters, ligatures not working, bad spacing etc). We will have a look and try to implement the changes in future versions.

---

## Background information

To help debugging, fill the following information before describing the issue:

* Font edition (i.e. trial, personal, plus, pro): Personal
* Font version (i.e. 1.805): 1.806
* The type of the font file you are using (i.e. otf, ttf): TTF
* Operating system (i.e. MacOS, Windows): Windows 10 21H1
* Display resolution (i.e. 3072x1920): 1920x1080
* Any other information that might be helpful: Sublime Text, Build 4126

## Issue

This is similar to spacing issues described in #19, #24, and #91, but it is still occurring for me. The triple-line ligatures are unequally spaced:

![image](https://user-images.githubusercontent.com/612027/155749727-4c1929c6-076f-4ebb-ab0a-10a6242a5b0a.png)

The degree of irregularity (and whether it's the upper or lower space that is larger) varies with font size; it occurs with most sizes, but goes away at certain "magic" sizes:

![image](https://user-images.githubusercontent.com/612027/155749588-395e939b-908a-4676-80dc-55f27d7172f5.png)

The first image was with the font size set to 12px; the second at 14px.

JetBrains Mono has the same problem. Cascadia Code doesn't (I tried a large variety of sizes with Cascadia Code, and every one was fine). Although not shown here, I see the same issues in VS Code.

**Comment by bebraw** [comment]
There's a technical limitation here. Especially on low resolution, it's possible to optimize the font only for specific pixel sizes. It's exactly this what you see in your screenshots. If we fix the issue for 12px, then it will likely break it for another size. That said, I am not sure if this is a problem we can fix as it's fundamentally not solvable.

I believe @MarcusSterz can provide more insight on this.

**Comment by Steve-OH** [comment]
I would accept that, were it not for the fact that, as I mentioned, the problem does not occur in Cascadia Code at all:

![image](https://user-images.githubusercontent.com/612027/155767050-c891b1b2-d760-49d2-be2e-dfd186a71f79.png)

Here's the source html for the above:
```html
<body>
  <div
    style="
      position: absolute;
      top: 10;
      left: 10;
      font-family: Cascadia Code;
      font-weight: 400;
    "
  >
    <h1>Cascadia Code</h1>
    <p style="font-size: 10px">fontSize === 10px;</p>
    <p style="font-size: 11px">fontSize === 11px;</p>
    <p style="font-size: 12px">fontSize === 12px;</p>
    <p style="font-size: 13px">fontSize === 13px;</p>
    <p style="font-size: 14px">fontSize === 14px;</p>
    <p style="font-size: 15px">fontSize === 15px;</p>
    <p style="font-size: 16px">fontSize === 16px;</p>
    <p style="font-size: 17px">fontSize === 17px;</p>
    <p style="font-size: 18px">fontSize === 18px;</p>
    <p style="font-size: 19px">fontSize === 19px;</p>
    <p style="font-size: 20px">fontSize === 20px;</p>
  </div>
  <div
    style="
      position: absolute;
      top: 10;
      left: 360;
      font-family: MonoLisa;
      font-weight: 400;
    "
  >
    <h1>MonoLisa</h1>
    <p style="font-size: 10px">fontSize === 10px;</p>
    <p style="font-size: 11px">fontSize === 11px;</p>
    <p style="font-size: 12px">fontSize === 12px;</p>
    <p style="font-size: 13px">fontSize === 13px;</p>
    <p style="font-size: 14px">fontSize === 14px;</p>
    <p style="font-size: 15px">fontSize === 15px;</p>
    <p style="font-size: 16px">fontSize === 16px;</p>
    <p style="font-size: 17px">fontSize === 17px;</p>
    <p style="font-size: 18px">fontSize === 18px;</p>
    <p style="font-size: 19px">fontSize === 19px;</p>
    <p style="font-size: 20px">fontSize === 20px;</p>
  </div>
</body>
```

**Comment by bebraw** [comment]
@Steve-OH Gotcha. Let's see if Marcus can figure this one out then. My understanding is that TTF might give some extra control over hinting for Windows so maybe they leveraged that.

**Comment by MarcusSterz** [comment]
@mekkablue Rainer, can you comment on that. If not let’s fix this, please!

**Comment by avidworks** [comment]
Just thought I'd chime in to say I just bought the latest version of this font, and am also having this issue on Windows at 1440p resolution. Screenshots are 18px font size at 400 and 600 weights however this issue occurs with all weights at various font sizes with differing results.  I compared to version 1.303 which did not seem to have the issue but is significantly blurrier (top screenshot). 

Font edition (i.e. trial, personal, plus, pro): Plus
Font version (i.e. 1.805): 1.808
The type of the font file you are using (i.e. otf, ttf): TTF
Operating system (i.e. MacOS, Windows): Windows 11
Display resolution (i.e. 3072x1920): 2560 x 1440
Any other information that might be helpful: Not application specific, this is just a hinting issue.

![image](https://user-images.githubusercontent.com/68607953/163751947-c94e4f82-000d-4083-ab3a-c94ea0024fe1.png)

![image](https://user-images.githubusercontent.com/68607953/163754189-3c8a9054-4c4e-4e42-8864-56bd0b43e61e.png)
![image](https://user-images.githubusercontent.com/68607953/163754204-f3312ece-d26e-4f28-8a90-4d5ce6cf4700.png)
![image](https://user-images.githubusercontent.com/68607953/163754210-72625940-1640-408f-903e-536404fb9930.png)
![image](https://user-images.githubusercontent.com/68607953/163754220-8cb15486-adb0-440d-92d0-13b09bc5abb7.png)


Like mentioned, Cascadia does not have this issue at any size, however it's worth pointing out that @aaronbell designed this font for use in Microsoft's OS, so a significant amount of time likely went into manually hinting each ligature. Almost every other font I've tried has this issue, so their case is certainly the exception. Still, it would be appreciated if looked into considering the price of this font.

**Comment by aaronbell** [comment]
IMO, well-hinted TTFs will always give better results on Windows than OTFs. There's simply more control. 

If y'all need a TTF hinter, you know where to find me :)

**Comment by MarcusSterz** [comment]
Meanwhile I have hinted the particular glyphs. Let’s see if it improves the rendering in the next version.

**Comment by bebraw** [comment]
@Steve-OH There's a 2.0 prerelease available from the orders page with the tweaks included. It would be cool if you could try it to see how hinting works out now.

**Comment by Steve-OH** [comment]
@bebraw It's definitely an improvement, although the results are slightly mixed:
![image](https://user-images.githubusercontent.com/612027/188476538-295222dd-8a7e-4dcc-9c83-055cbd3b461c.png)

The === is much better, but at the expense of the single = (end of first line) and to a lesser extent the =>.

And while the spacing between the lines is good at all sizes, the line weights are a bit uneven at some sizes:
![image](https://user-images.githubusercontent.com/612027/188476999-b671404f-257f-4428-a787-42b763503d8b.png)

Overall, I think I'd rate this version over the previous one.

**Comment by Steve-OH** [comment]
@bebraw Here's another oddity with the new font, possibly worthy of a new issue (I don't think it was there with the earlier version). At some font sizes, the left and right parentheses, square brackets, and curly braces are vertically misaligned with respect to each other (see especially 12px and also 18px, where the offset is in the opposite direction):
![image](https://user-images.githubusercontent.com/612027/188516425-662fd364-cbff-4fbf-ab30-ad0369b18055.png)

I see the same behavior at https://www.monolisa.dev/specimen, in both Chrome and Firefox.

**Comment by MarcusSterz** [comment]
Regarding === this issue is almost impossible to prevent since we are dealing with a so-called moiré effect: Two grids overlay each other and form an unwanted, hardly predictable result. 

Looking into the bracket thing though ...

**Comment by aaronbell** [comment]
If I had to guess, one bracket is a horizontally mirrored & vertically flipped component of the other. So the hints are upside down and under certain conditions rounding occurs differently. 

Changing it to just mirrored would solve it.

**Comment by MarcusSterz** [comment]
> If I had to guess, one bracket is a horizontally mirrored & vertically flipped component of the other. So the hints are upside down and under certain conditions rounding occurs differently.
> 
> Changing it to just mirrored would solve it.

That’s very likley the case.

**Comment by bebraw** [comment]
This has been addressed in newly published 2.0.
